dask/array/_shuffle.py

```diff
@@ -0,0 +1,199 @@
+from __future__ import annotations
+
+from itertools import count, product
+
+import numpy as np
+
+from dask import config
+from dask.array.chunk import getitem
+from dask.array.core import Array, unknown_chunk_message
+from dask.base import tokenize
+from dask.highlevelgraph import HighLevelGraph
+
+
+def shuffle(x, indexer: list[list[int]], axis):
+    """
+    Reorders one dimensions of a Dask Array based on an indexer.
+
+    The indexer defines a list of positional groups that will end up in the same chunk
+    together. A single group is in at most one chunk on this dimension, but a chunk
+    might contain multiple groups to avoid fragmentation of the array.
+
+    The algorithm tries to balance the chunksizes as much as possible to ideally keep the
+    number of chunks consistent or at least manageable.
+
+    Parameters
+    ----------
+    x: dask array
+        Array to be shuffled.
+    indexer:  list[list[int]]
```

Comment:
nit: It might be safer to convert the indexer to a `tuple[tuple[int]]` internally to avoid unintentional modification.

Reply:
I am messing with list methods later, do you want to protect against modifying the input?

Reply:
> do you want to protect against modifying the input?

Yes, I was wondering if that would be helpful here.

Reply:
I would rather copy in this case, you ok with?

Reply:
sounds good!